File: ruby/pairing.rb
class Node
	attr_accessor :key, :children

	def initialize(key)
		@key, @children = key, []
	end

	def link(that)
		if that.nil? then
			return self
		elsif key > that.key then
			return that.link(self)
		else
			children << that
			return self
		end
	end

	def self.merge_list(children)
		if children.empty? then
			nil
		elsif children.length == 1 then
			children[0]
		else
			rem = Node.merge_list(children[2..-1])
			return children[0].link(children[1]).link(rem)
		end
	end

	def merged_children
		new = Node.merge_list(children.reverse)
		children = []
		return new
	end
end

class Heap
	@root = nil

	def top
		return @root.key
	end

	def pop
		@root = @root.merged_children
	end

	def insert(key)
		@root = Node.new(key).link(@root)
	end
end

heap = Heap.new

STDIN.read.split("\n").each do |line|
	words = line.split(" ")

	if words[0] == "0" then
		heap.insert(words[1].to_i)
	elsif words[0] == '1' then
		puts heap.top
	else
		heap.pop
	end
end

File: python/pairing.py
import sys

class Node():
    """ Pairing heap node """

    def __init__(self, key):
        self.key = key
        self.children = []

    def link(self, other):
        if other == None:
            return self

        if self.key > other.key:
            return other.link(self)

        self.children.append(other)
        return self

    @classmethod
    def merge_list(cls, children):
        if children == []:
            return None

        if len(children) == 1:
            return children[0]

        rem = cls.merge_list(children[2:])

        return children[0].link(children[1]).link(rem)

    def merged_children(self):
        self.children.reverse()
        new = self.merge_list(self.children)
        self.children = []
        return new

class Heap():
    def __init__(self):
        self.root = None

    def top(self):
        return self.root.key

    def pop(self):
        self.root = self.root.merged_children()

    def insert(self, key):
        self.root = Node(key).link(self.root)

if __name__ == '__main__':
    heap = Heap()

    for line in sys.stdin:
        data = line.split()
        if data[0] == "0":
            heap.insert(int(data[1]))
        elif data[0] == "1":
            print(heap.top())
        else:
            heap.pop()

